frontend/persist.lua

```diff
@@ -0,0 +1,131 @@
+local bitser = require("ffi/bitser")
+local dump = require("dump")
+local lfs = require("libs/libkoreader-lfs")
+
+local codecs = {
+    -- bitser: binary form, fast encode/decode, low size. Not human readable.
+    bitser = {
+
+        serialize = function(t, file)
+            local ok, str = pcall(bitser.dumps, t)
+            if not ok then
+                return nil, "cannot serialize " .. tostring(t)
+            end
+            return str
+        end,
+
+        deserialize = function(file)
+            local f, err, str
+            f, err = io.open(file, "rb")
+            if not f then
+                return nil, err
+            end
+            str, err = f:read("*a")
+            f:close()
+            if not str then
+                return nil, err
+            end
+            local ok, t = pcall(bitser.loads, str)
+            if not ok then
+                return nil, "malformed serialized data"
+            end
+            return t
+        end,
+    },
+    -- dump: human readable, pretty printed, fast enough for most user cases.
+    dump = {
+
+        serialize = function(t, file, as_bytecode)
+            local content, err = dump(t)
+            if not content then
+                return nil, string.format("cannot serialize table %s: %s", t, err)
+            end
+            local str
+            if as_bytecode then
+                str, err =  load("return " .. content)
+                if not str then
+                    print("cannot convert table to bytecode: %s, ignoring", err)
+                else
+                    str = string.dump(str, true)
+                end
+            end
+
+            if not str then
+                str = "return " .. content
+            end
+            return str
+        end,
+
+        deserialize = function(file)
+            local ok, t, err = pcall(dofile, file)
+            if not ok then
+                return nil, err
+            end
+            return t
+        end,
+    }
+}
+
+local Persist = {}
+
+function Persist:new(o)
+    o = o or {}
+    setmetatable(o, self)
+    self.__index = self
+    return o:init(o.path, o.codec)
+end
+
+function Persist:init(path, codec)
+    if type(path) ~= "string" then
+        return nil, "path is required"
```

Comment:
Shouldn't this just crash? It's a coding error if we don't provide a string, or are there other use cases?
Also not sure what happens when :init() returns nil :) does :new() still return something or something usable?

Reply:
```lua
local some_cache = Persist:new{ codec = "dump" }
```

will return nil instead of a table. Maybe it is better an assert and move the init stuff to new?, I'm a bit fanatic of `something, err = SomeClass:new{}` but maybe doesn't make much sense here, as we don't "catch" errors :p

Reply:
> a bit fanatic of something, err = SomeClass:new{}

Do we have that kind of thing ? Feels like I've never seen them :)

Reply:
Ummh, nope :P.

Went the assert way and removed the init method, which was confusing.